### ISSUE TITLE
Ensemble model training/running with different architectures

### ISSUE THREAD

**Issue by PeishiJiang** [patrick-kidger/equinox#786]
Hi Equinox users and developers,

I want to use ```vmap``` to parallelize the training/execution of ```eqx.Module```-based models. The ensemble example provided [here](https://docs.kidger.site/equinox/tricks/#ensembling) only works for models with the same architecture (but different initializations).

Is there a way/an example for ```vmap```ing the models with varying architectures? It would be very useful for hyperparameter tuning, in which case we often need to tune the number of layers/widths/etc.

I have failed in multiple attempts, including the try below (```vmap``` only working for ```jnp.array```, but not a list of pytrees):

```python
import jax
import jax.random as jrandom

import equinox as eqx

key = jrandom.key(1111)

# Generate multiple MLP models with different depths
depth_ens = list(range(1,10))
model_ens = [eqx.nn.MLP(2, 2, 2, depth=depth, key=key) for depth in depth_ens]

# Provide some input data
inputs = jrandom.uniform(key, shape=(2,))

# Run one single model
def run_model(model, inputs):
    return model(inputs)
# run_model(model_ens[-1], inputs)

# Run model ensemble
run_model_ens = eqx.filter_vmap(run_model, in_axes=(0, None))
run_model_ens(model_ens, inputs)

```

Thanks!
Peishi

**Comment by lockwo** [comment]
IMO this is sort of a fundamental divergence from what jax does. I've thought about this some too. Since in the case of this vmap-ing, equinox isn't really doing anything fancy/important, we can just reduce it to jax. Basically, what you want to do, is vmap over pytrees of different shapes/structures. Like if depth is a list of networks, the pytreedef of different depths is different. Jax has a hard time working with this because vmap just wants to chop off the first dimension of axes, but these don't share the same axes. This can reduce down to a common problem I've seen before, which is you want to vmap over "arrays of pytrees" but jax needs "pytrees of arrays". There are a couple solutions you might consider, the most trivial of which is just padding out smaller pytrees to match the shapes of everything else.

**Comment by PeishiJiang** [comment]
> IMO this is sort of a fundamental divergence from what jax does. I've thought about this some too. Since in the case of this vmap-ing, equinox isn't really doing anything fancy/important, we can just reduce it to jax. Basically, what you want to do, is vmap over pytrees of different shapes/structures. Like if depth is a list of networks, the pytreedef of different depths is different. Jax has a hard time working with this because vmap just wants to chop off the first dimension of axes, but these don't share the same axes. This can reduce down to a common problem I've seen before, which is you want to vmap over "arrays of pytrees" but jax needs "pytrees of arrays". There are a couple solutions you might consider, the most trivial of which is just padding out smaller pytrees to match the shapes of everything else.

Thank you for the replies, @lockwo ! 

I have thought about the padding option but figured it might be a bit too trivial and does not sound a 'perfect' solution. What are the other options? 

Or is there another way, that goes beyond vmap, to parallelize the ensemble training. I've seen/used [ray tune](https://docs.ray.io/en/latest/tune/tutorials/tune-resources.html#how-to-leverage-gpus-in-tune) to do the same stuff for pytorch models. However, ray tune currently does not directly support JAX. I'm wondering whether there is a similar tool for JAX out there.

**Comment by patrick-kidger** [comment]
If you want to distribute parallel training of an ensemble then you have loads of options. Do each model in a separate Python process and parallelise at that level.

If you need to do stuff at the JAX level then the closest you'll get is `jax.lax.axis_index` + `jax.lax.switch`, although I have no guarantee that will work well. At the very least I expect you should put that above any `grad`, `vmap` transforms etc. I think that's firmly in 'if you know what you're doing' territory.

**Comment by PeishiJiang** [comment]
> If you want to distribute parallel training of an ensemble then you have loads of options. Do each model in a separate Python process and parallelise at that level.
> 
> If you need to do stuff at the JAX level then the closest you'll get is `jax.lax.axis_index` + `jax.lax.switch`, although I have no guarantee that will work well. At the very least I expect you should put that above any `grad`, `vmap` transforms etc. I think that's firmly in 'if you know what you're doing' territory.

Thanks, Patrick! My follow-up question would be: How do I designate a fraction of a GPU to train each equinox model? (Doing it through multiprocessing in CPU is more straightforward).

Say I want to train multiple small MLPs with different hyperparameters. Sequentially training on one GPU is inefficient. It would be great if I could allocate a proportion of GPU (e.g., 20% for each model) to train each MLP. Is there a convenient JAX tool in this regard?

I figure this question is probably beyond the equinox. So, feel free to close the comment if needed. But any feedback is welcome.

**Comment by patrick-kidger** [comment]
Indeed this isn't really an Equinox thing!
But regardless -- if they're all part of the same JAX computation graph then that should happen automatically. If they're not then you'll probably need to set the reserved memory usage:

https://jax.readthedocs.io/en/latest/gpu_memory_allocation.html

but that otherwise things should continue to just work.

**Comment by PeishiJiang** [comment]
> Indeed this isn't really an Equinox thing! But regardless -- if they're all part of the same JAX computation graph then that should happen automatically. If they're not then you'll probably need to set the reserved memory usage:
> 
> https://jax.readthedocs.io/en/latest/gpu_memory_allocation.html
> 
> but that otherwise things should continue to just work.

Thank you! I'll close the issue.

I'm using joblib for multiple threads processing on the CPU for now. 

For GPU, XLA_PYTHON_CLIENT_MEM_FRACTION seems to only allocate GPU to an 'overall' Python execution environment. I'm not sure how it can be used to specify the usages for multiple models unless these models are initialized and called in separate Python executions. Let me know if I missed anything.
